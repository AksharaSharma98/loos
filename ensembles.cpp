/*
  This file is part of LOOS.

  LOOS (Lightweight Object-Oriented Structure library)
  Copyright (c) 2008, Tod D. Romo, Alan Grossfield
  Department of Biochemistry and Biophysics
  School of Medicine & Dentistry, University of Rochester

  This package (LOOS) is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation under version 3 of the License.

  This package is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/





#include <ensembles.hpp>
#include <XForm.hpp>
#include <AtomicGroup.hpp>
#include <Trajectory.hpp>

namespace loos {

  // Assume all groups are already sorted or matched...

  AtomicGroup averageStructure(const std::vector<AtomicGroup>& ensemble) {
    AtomicGroup avg = ensemble[0].copy();

    // First, zap our coords...
    int n = avg.size();
    int i;
    for (i=0; i<n; i++)
      avg[i]->coords() = GCoord(0.0, 0.0, 0.0);

    // Now, accumulate...
    std::vector<AtomicGroup>::const_iterator j;
    for (j = ensemble.begin(); j != ensemble.end(); j++) {
      for (i = 0; i<n; i++)
        avg[i]->coords() += (*j)[i]->coords();
    }

    for (i=0; i<n; i++)
      avg[i]->coords() /= ensemble.size();

    return(avg);
  }



  AtomicGroup averageStructure(const std::vector<AtomicGroup>& ensemble, const std::vector<XForm>& xforms) {
    if (xforms.size() != ensemble.size())
      throw(LOOSError("Transforms do not match the passed ensemble in loos::averageStructure()"));

    AtomicGroup avg = ensemble[0].copy();

    // First, zap our coords...
    int n = avg.size();
    int i;
    for (i=0; i<n; i++)
      avg[i]->coords() = GCoord(0.0, 0.0, 0.0);

    // Now, accumulate...
    for (uint j=0; j<ensemble.size(); ++j) {
      AtomicGroup frame = ensemble[j].copy();
      frame.applyTransform(xforms[j]);
      for (i = 0; i<n; i++)
        avg[i]->coords() += frame[i]->coords();
    }
    
    for (i=0; i<n; i++)
      avg[i]->coords() /= ensemble.size();

    return(avg);
  }




  AtomicGroup averageStructure(const AtomicGroup& g, const std::vector<XForm>& xforms, pTraj& traj, std::vector<uint>& frame_indices) {
    AtomicGroup avg = g.copy();
    AtomicGroup frame = g.copy();
    int n = avg.size();
    for (int i=0; i<n; i++)
      avg[i]->coords() = GCoord(0.0, 0.0, 0.0);
    
    uint tn = traj->nframes();
    uint fn = frame_indices.size();
    
    // Safety check...
    if (fn != xforms.size())
      throw(std::runtime_error("Mismatch in number of frames in the trajectory requested and passed transforms for loos::averageStructure()"));
    
    for (uint j=0; j<fn; ++j) {
      if (frame_indices[j] >= tn)
        throw(std::runtime_error("Frame index exceeds trajectory size"));
      
      traj->readFrame(frame_indices[j]);
      traj->updateGroupCoords(frame);
      frame.applyTransform(xforms[j]);
      for (int i=0; i<n; i++)
        avg[i]->coords() += frame[i]->coords();
    }
    
    for (int i=0; i<n; i++)
      avg[i]->coords() /= fn;
    
    return(avg);
  }
  
  
  
  AtomicGroup averageStructure(const AtomicGroup& g, const std::vector<XForm>& xforms, pTraj& traj) {
    uint nx = xforms.size();
    uint nt = traj->nframes();
    
    if (nt != nx)
      throw(std::runtime_error("Mismatch in number of frames in the trajectory and passed transforms"));
    std::vector<uint> frame_indices;
    for (uint i=0; i<nt; ++i)
      frame_indices.push_back(i);

    return(averageStructure(g, xforms, traj, frame_indices));
  }





  boost::tuple<std::vector<XForm>,greal,int> iterativeAlignment(std::vector<AtomicGroup>& ensemble,
                                                                greal threshold, int maxiter) {
    int iter = 0;
    int n = ensemble.size();
    greal rms;
    std::vector<XForm> xforms(n);
    AtomicGroup avg;

    // Start by aligning against the first structure in the ensemble
    AtomicGroup target = ensemble[0].copy();

    target.centerAtOrigin();

    do {
      for (int i = 0; i<n; i++) {
        GMatrix M = ensemble[i].alignOnto(target);
        xforms[i].premult(M);
      }

      avg = averageStructure(ensemble);
      rms = avg.rmsd(target);
      target = avg;
      iter++;
    } while (rms > threshold && iter <= maxiter );
    
    boost::tuple<std::vector<XForm>, greal, int> res(xforms, rms, iter);
    return(res);
  }



  boost::tuple<std::vector<XForm>, greal, int> iterativeAlignment(const AtomicGroup& g, pTraj& traj, std::vector<uint>& frame_indices,
                                                                  greal threshold, int maxiter) {
    std::vector<AtomicGroup> frames;

    uint nf = frame_indices.size();
    uint nt = traj->nframes();

    for (uint i=0; i<nf; ++i) {
      if (frame_indices[i] >= nt)
        throw(std::runtime_error("Frame index exceeds trajectory size"));

      traj->readFrame(frame_indices[i]);
      AtomicGroup frame = g.copy();
      traj->updateGroupCoords(frame);
      frames.push_back(frame);
    }

    boost::tuple<std::vector<XForm>, greal, int> result = iterativeAlignment(frames, threshold, maxiter);
    return(result);
  }


  boost::tuple<std::vector<XForm>, greal, int> iterativeAlignment(const AtomicGroup& g, pTraj& traj,
                                                                  greal threshold, int maxiter) {

    std::vector<uint> frame_indices;
    uint nt = traj->nframes();
    for (uint i=0; i<nt; ++i)
      frame_indices.push_back(i);
    return(iterativeAlignment(g, traj, frame_indices, threshold, maxiter));
  }



  void applyTransforms(std::vector<AtomicGroup>& ensemble, std::vector<XForm>& xforms) {
    uint n = ensemble.size();
    if (n != xforms.size())
      throw(std::runtime_error("Mismatch in the size of the ensemble and the transformations"));

    for (uint i=0; i<n; ++i)
      ensemble[i].applyTransform(xforms[i]);
  }



  void readTrajectory(std::vector<AtomicGroup>& ensemble, const AtomicGroup& model, pTraj trajectory) {
    AtomicGroup clone = model.copy();
    
    while (trajectory->readFrame()) {
      trajectory->updateGroupCoords(clone);
      AtomicGroup frame = clone.copy();
      ensemble.push_back(frame);
    }
  }


  void readTrajectory(std::vector<AtomicGroup>& ensemble, const AtomicGroup& model, pTraj trajectory, std::vector<uint>& frames) {
    AtomicGroup clone = model.copy();
    
    std::vector<uint>::iterator i;
    for (i = frames.begin(); i != frames.end(); ++i) {
      if (*i >= trajectory->nframes())
        throw(std::runtime_error("Frame index exceeds trajectory size in readTrajectory()"));
      trajectory->readFrame(*i);
      trajectory->updateGroupCoords(clone);
      AtomicGroup frame = clone.copy();
      ensemble.push_back(frame);
    }
  }



  RealMatrix extractCoords(std::vector<AtomicGroup>& ensemble) {
    uint n = ensemble.size();
    uint m = ensemble[0].size();
    RealMatrix M(3*m, n);

    for (uint i=0; i<n; ++i)
      for (uint j=0; j<m; ++j) {
        M(3*j, i) = ensemble[i][j]->coords().x();
        M(3*j+1, i) = ensemble[i][j]->coords().y();
        M(3*j+2, i) = ensemble[i][j]->coords().z();
      }

    return(M);
  }


  RealMatrix extractCoords(std::vector<AtomicGroup>& ensemble, std::vector<XForm>& xforms) {
    uint n = ensemble.size();

    if (n != xforms.size())
      throw(std::runtime_error("Mismatch between the size of the ensemble and the transformations"));

    uint m = ensemble[0].size();
    RealMatrix M(3*m, n);

    for (uint i=0; i<n; ++i) {
      GMatrix W = xforms[i].current();

      for (uint j=0; j<m; ++j) {
        GCoord c = W * ensemble[i][j]->coords();
        M(3*j, i) = c.x();
        M(3*j+1, i) = c.y();
        M(3*j+2, i) = c.z();
      }
    }

    return(M);
  }


  void subtractAverage(RealMatrix& M) {
    uint m = M.rows();
    uint n = M.cols();
    double *avg = new double[m];
    
    for (uint j=0; j<m; ++j)
      avg[j] = 0.0;

    for (uint i=0; i<n; ++i)
      for (uint j=0; j<m; ++j)
        avg[j] += M(j, i);

    for (uint j=0; j<m; ++j)
      avg[j] /= n;

    for (uint i=0; i<n; ++i)
      for (uint j=0; j<m; ++j)
        M(j,i) -= avg[j];
  }


  boost::tuple<RealMatrix, RealMatrix, RealMatrix> svd(std::vector<AtomicGroup>& ensemble, bool align) {

    RealMatrix M;
    if (align) {
      boost::tuple<std::vector<XForm>, greal, int> res = iterativeAlignment(ensemble);
      M = extractCoords(ensemble, boost::get<0>(res));
    } else
      M = extractCoords(ensemble);

    subtractAverage(M);
    boost::tuple<RealMatrix, RealMatrix, RealMatrix> res = Math::svd(M);
    return(res);
  }



}
