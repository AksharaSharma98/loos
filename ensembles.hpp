/*
  This file is part of LOOS.

  LOOS (Lightweight Object-Oriented Structure library)
  Copyright (c) 2008, Tod D. Romo, Alan Grossfield
  Department of Biochemistry and Biophysics
  School of Medicine & Dentistry, University of Rochester

  This package (LOOS) is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation under version 3 of the License.

  This package is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/



#if !defined(ENSEMBLES_HPP)
#define ENSEMBLES_HPP

#include <vector>
#include <XForm.hpp>

#include <loos_defs.hpp>

namespace loos {
  //! Compute the average structure of a set of AtomicGroup objects
  AtomicGroup averageStructure(const vector<AtomicGroup>& ensemble);

  //! Compute the average structure by reading through a trajectory
  AtomicGroup averageStructure(const AtomicGroup&, const vector<XForm>&, pTraj);

  //! Compute an iterative superposition (a la Alan)
  boost::tuple<vector<XForm>, greal, int> iterativeAlignment(vector<AtomicGroup>& ensemble, greal threshold, int maxiter=1000);

  //! Compute an iterative superposition by reading in frames from the Trajectory.
  /*!
   * This function will internally cache an AtomicGroup copy for each
   * frame of the trajectory.  This could chew up a lot of memory, but
   * we make the assumption that you will usually be aligning against
   * a fairly small subset of each frame...
   */
  boost::tuple<vector<XForm>, greal, int> iterativeAlignment(const AtomicGroup& g, pTraj, greal threshold, int maxiter=1000);
};



#endif
