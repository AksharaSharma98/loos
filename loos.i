%module loos


%feature("autodoc", "1");

%include <std_string.i>
%include <std_vector.i>
%include <boost_shared_ptr.i>

typedef unsigned int    uint;
typedef unsigned long     ulong;

namespace loos {
  typedef double greal;
  typedef long gint;

  typedef float dcd_real;
  typedef double dcd_double;

  typedef Coord<double> GCoord;
}

%include "Coord.i"
%include "Atom.i"
%include "Matrix44.i"
%include "pdb_remarks.i"
%include "XForm.i"
%include "AtomicGroup.i"
%include "Trajectory.i"
%include "utils.i"
%include "cryst.i"
%include "pdb.i"
%include "sfactories.i"
%include "dcdwriter.i"
%include "ensembles.i"
%include "Geometry.i"
%include "TimeSeries.i"
%include "HBondDetector.i"
%include "exceptions.i"
