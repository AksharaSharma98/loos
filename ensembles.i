/*
  This file is part of LOOS.

  LOOS (Lightweight Object-Oriented Structure library)
  Copyright (c) 2008, Tod D. Romo, Alan Grossfield
  Department of Biochemistry and Biophysics
  School of Medicine & Dentistry, University of Rochester

  This package (LOOS) is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation under version 3 of the License.

  This package is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/



%header %{
#include <ensembles.hpp>
%}

namespace loos {

  AtomicGroup averageStructure(const std::vector<AtomicGroup>& ensemble);

  AtomicGroup averageStructure(const std::vector<AtomicGroup>& ensemble, const std::vector<XForm>& xforms);

  AtomicGroup averageStructure(const AtomicGroup&, const std::vector<XForm>&, pTraj& traj, const std::vector<uint>& indices);

  AtomicGroup averageStructure(const AtomicGroup&, const std::vector<XForm>&, pTraj& traj);

  boost::tuple<std::vector<XForm>, greal, int> iterativeAlignment(std::vector<AtomicGroup>& ensemble, greal threshold = 1e-6, int maxiter=1000);

  boost::tuple<std::vector<XForm>, greal, int> iterativeAlignment(const AtomicGroup& g, pTraj& traj, const std::vector<uint>& frame_indices, greal threshold = 1e-6, int maxiter=1000);

  boost::tuple<std::vector<XForm>, greal, int> iterativeAlignment(const AtomicGroup& g, pTraj& traj, greal threshold = 1e-6, int maxiter=1000);


  void applyTransforms(std::vector<AtomicGroup>& ensemble, std::vector<XForm>& xforms);

  void readTrajectory(std::vector<AtomicGroup>& ensemble, const AtomicGroup& model, pTraj trajectory);
  void readTrajectory(std::vector<AtomicGroup>& ensemble, const AtomicGroup& model, pTraj trajectory, std::vector<uint>& frames);
};

