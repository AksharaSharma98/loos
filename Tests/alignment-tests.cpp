/*
  alignment-tests.cpp
  (c) 2008 Tod D. Romo


  Grossfield Lab
  Department of Biochemistry and Biophysics
  University of Rochester Medical School

  Unit testing for alignment routines...

*/

#include <iostream>
#include <iomanip>
#include <boost/random.hpp>
#include <ctime>


#include <loos.hpp>
#include <pdb.hpp>
#include <Kernel.hpp>
#include <Parser.hpp>
#include <Selectors.hpp>
#include <ensembles.hpp>

// Configuration options...
const unsigned int iter_tests = 1000;
const double iter_perturbation = 2.0;
const double iter_rmsd_thresh = 2.0;
const double iter_final_rmsd_thresh = 1e-3;

const unsigned int single_tests = 1000;
const double single_rmsd_thresh = 1e-10;
const static bool show_results = false;



int main(int argc, char *argv[]) {
  unsigned int i;

  if (argc != 2) {
    cerr << "Usage- " << argv[0] << " pdbfile\n";
    exit(-1);
  }
  
  // Suppress for easy diffs...
  //cout << invocationHeader(argc, argv) << endl;

  loos::base_generator_type& rng = loos::rng_singleton();

  // Uncommont the follong to seed the suite-wide RNG
  //loos::randomSeedRNG();


  PDB pdb(argv[1]);

  CAlphaSelector casel;
  AtomicGroup cas = pdb.select(casel);
  cout << "Found " << cas.size() << " CAlphas.\n";

  AtomicGroup casb = cas.copy();
  casb.perturbCoords(1.0);
  greal rmsd = cas.rmsd(casb);
  cout << "RMSD test = " << rmsd << endl;
  XForm W;
  W.rotate('Y', 35.0);
  casb.applyTransform(W);
  rmsd = cas.rmsd(casb);
  cout << "Rotated rmsd = " << rmsd << endl;

  GMatrix M = casb.alignOnto(cas);
  cout << "Aligned rmsd = " << cas.rmsd(casb) << endl;
  cout << M << endl;


  boost::uniform_real<> anglemap(-180.0, 180.0);
  boost::variate_generator<loos::base_generator_type&, boost::uniform_real<> > angles(rng, anglemap);

  boost::uniform_real<> transmap(-20.0, 20.0);
  boost::variate_generator<loos::base_generator_type&, boost::uniform_real<> > translations(rng, transmap);


  // -------------------------------------------------------------------------------

  cout << "*** Single Alignment Tests ***\n";

  int warnings = 0;
  for (i=0; i<single_tests; i++) {
    AtomicGroup casr = cas.copy();
    W.identity();
    W.translate(translations(), translations(), translations());
    W.rotate('x', angles());
    W.rotate('y', angles());
    W.rotate('z', angles());
    casr.applyTransform(W);
    greal pre_rmsd = cas.rmsd(casr);
    casr.alignOnto(cas);
    double rmsd = cas.rmsd(casr);
    if (rmsd >= single_rmsd_thresh) {
      cout << "WARNING - Possible mis-alignment - pre = " << pre_rmsd << ", post = " << rmsd << endl;
      ++warnings;
    }
  }

  if (warnings > 0)
    cout << "*** There were " << warnings << " possible errors detected.\n";
  else
    cout << "All tests passed (threshold = " << single_rmsd_thresh << ")\n";



  // -------------------------------------------------------------------------------
  // Now run iteratative superpositon tests...

  vector<AtomicGroup> mols;
  for (i=0; i<iter_tests; i++) {
    AtomicGroup subgroup = cas.copy();
    GCoord t(translations(), translations(), translations());
    
    subgroup.perturbCoords(iter_perturbation);

    W.identity();
    W.translate(t);
    W.rotate('z', angles());
    W.rotate('y', angles());
    W.rotate('x', angles());
	     
    subgroup.applyTransform(W);
    mols.push_back(subgroup);
  }

  AtomicGroup avg = averageStructure(mols);
  if (show_results) {
    cout << "Pre-aligned rmsds:\n";
    for (i=0; i<iter_tests; i++)
      cout << "\t" << i << "\t" << avg.rmsd(mols[i]) << endl;
  }


  greal final_rmsd = loos::iterativeAlignment(mols, 2.0);
  if (show_results)
    cout << "Final alignment rmsd to avg struct = " << final_rmsd << endl;

  if (final_rmsd >= iter_final_rmsd_thresh)
    cout << "WARNING - final rmsd of " << final_rmsd << " is above threshold.\n";

  warnings = 0;
  avg = averageStructure(mols);
  for (i=0; i<iter_tests; i++) {
    if (show_results)
      cout << "\t" << i << "\t" << avg.rmsd(mols[i]) << endl;
    double irmsd = avg.rmsd(mols[i]);
    if (irmsd >= iter_rmsd_thresh) {
      ++warnings;
      cout << "WARNING - possible iterative failure at " << i << " with rmsd of " << irmsd << endl;
    }
  }

  if (warnings > 0)
    cout << "*** There were " << warnings << " possible errors detected.\n";
  else
    cout << "All tests passed (threshold = " << iter_final_rmsd_thresh << " : " << iter_rmsd_thresh << ")\n";
}
