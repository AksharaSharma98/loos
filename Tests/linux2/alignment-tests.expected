

-=-=-=-=-=-=-=- MACROCYCLE #0 -=-=-=-=-=-=-=-
**************************************************
Selected 28 atoms with selector 'name == 'CA''.
RMSD test = 1
Rotated rmsd = 36.2336
Aligned rmsd = 0.942671
[0.801764 0.00393839 -0.597628 1.26605;0.0130238 0.999626 0.02406 -1.09837;0.597499 -0.0270739 0.801412 0.978502;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 4
WARNING - possible iterative failure at 245 with rmsd of 2.00194
WARNING - possible iterative failure at 390 with rmsd of 2.00154
WARNING - possible iterative failure at 549 with rmsd of 2.00114
WARNING - possible iterative failure at 689 with rmsd of 2.00077
WARNING - possible iterative failure at 727 with rmsd of 2.00302
WARNING - possible iterative failure at 831 with rmsd of 2.00149
*** There were 6 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************
**************************************************
Selected 25 atoms with selector 'resid == 5'.
RMSD test = 1
Rotated rmsd = 22.0809
Aligned rmsd = 0.856746
[0.833129 0.128445 -0.537958 -2.44657;-0.000156661 0.972714 0.232006 1.86217;0.553079 -0.193206 0.810416 4.36095;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 6
WARNING - possible iterative failure at 319 with rmsd of 2.00585
WARNING - possible iterative failure at 718 with rmsd of 2.00019
*** There were 2 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************
**************************************************
Selected 125 atoms with selector 'segid == 'L2''.
RMSD test = 1
Rotated rmsd = 17.4178
Aligned rmsd = 0.990014
[0.811883 -0.00177016 -0.583818 -0.072647;-0.00128186 0.999988 -0.00481461 -0.0798121;0.583819 0.00465727 0.811871 0.164459;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 4
WARNING - possible iterative failure at 561 with rmsd of 2.00167
WARNING - possible iterative failure at 679 with rmsd of 2.00312
WARNING - possible iterative failure at 891 with rmsd of 2.00063
WARNING - possible iterative failure at 936 with rmsd of 2.00022
*** There were 4 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************


-=-=-=-=-=-=-=- MACROCYCLE #1 -=-=-=-=-=-=-=-
**************************************************
Selected 28 atoms with selector 'name == 'CA''.
RMSD test = 1
Rotated rmsd = 35.8753
Aligned rmsd = 0.915275
[0.820921 -0.0166791 -0.570797 1.35804;0.00968673 0.999836 -0.0152844 -0.0948567;0.570959 0.00701812 0.820949 -0.528579;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 4
WARNING - possible iterative failure at 317 with rmsd of 2.00351
WARNING - possible iterative failure at 958 with rmsd of 2.00147
*** There were 2 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************
**************************************************
Selected 25 atoms with selector 'resid == 5'.
RMSD test = 1
Rotated rmsd = 22.1802
Aligned rmsd = 0.91609
[0.776867 -0.0865219 -0.623692 0.366318;0.0388701 0.995215 -0.0896453 0.728485;0.628464 0.0453995 0.776513 0.58165;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 6
WARNING - possible iterative failure at 147 with rmsd of 2.00169
WARNING - possible iterative failure at 814 with rmsd of 2.00019
WARNING - possible iterative failure at 899 with rmsd of 2.00152
*** There were 3 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************
**************************************************
Selected 125 atoms with selector 'segid == 'L2''.
RMSD test = 1
Rotated rmsd = 17.2577
Aligned rmsd = 0.994444
[0.817325 -0.000168311 -0.576177 -0.110243;-0.0042518 0.999971 -0.00632341 -0.101071;0.576161 0.00761807 0.8173 -0.0789947;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 4
WARNING - possible iterative failure at 10 with rmsd of 2.00099
WARNING - possible iterative failure at 45 with rmsd of 2.00042
WARNING - possible iterative failure at 268 with rmsd of 2.00227
WARNING - possible iterative failure at 420 with rmsd of 2.00182
WARNING - possible iterative failure at 431 with rmsd of 2.00061
*** There were 5 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************


-=-=-=-=-=-=-=- MACROCYCLE #2 -=-=-=-=-=-=-=-
**************************************************
Selected 28 atoms with selector 'name == 'CA''.
RMSD test = 1
Rotated rmsd = 36.1642
Aligned rmsd = 0.977082
[0.819476 -0.0104075 -0.573018 0.747067;0.00253513 0.999891 -0.0145351 0.208125;0.573107 0.0104585 0.819414 -0.88854;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 4
WARNING - possible iterative failure at 689 with rmsd of 2.0044
WARNING - possible iterative failure at 738 with rmsd of 2.00576
WARNING - possible iterative failure at 769 with rmsd of 2.00378
WARNING - possible iterative failure at 960 with rmsd of 2.00489
*** There were 4 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************
**************************************************
Selected 25 atoms with selector 'resid == 5'.
RMSD test = 1
Rotated rmsd = 22.2126
Aligned rmsd = 0.97595
[0.781216 -0.0347744 -0.623291 -0.754393;0.0337215 0.99934 -0.0134892 1.1142;0.623349 -0.0104804 0.781874 1.82634;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 5
WARNING - possible iterative failure at 67 with rmsd of 2.00638
*** There were 1 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************
**************************************************
Selected 125 atoms with selector 'segid == 'L2''.
RMSD test = 1
Rotated rmsd = 17.3739
Aligned rmsd = 0.994423
[0.816435 -0.00251626 -0.577432 -0.0276081;0.00305574 0.999995 -3.71325e-05 0.103171;0.577429 -0.00173416 0.816439 0.0553112;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 4
WARNING - possible iterative failure at 15 with rmsd of 2.00142
WARNING - possible iterative failure at 66 with rmsd of 2.00261
WARNING - possible iterative failure at 108 with rmsd of 2.00281
WARNING - possible iterative failure at 189 with rmsd of 2.00001
WARNING - possible iterative failure at 272 with rmsd of 2.00179
WARNING - possible iterative failure at 365 with rmsd of 2.00008
WARNING - possible iterative failure at 405 with rmsd of 2.00636
WARNING - possible iterative failure at 406 with rmsd of 2.00081
WARNING - possible iterative failure at 409 with rmsd of 2.00418
WARNING - possible iterative failure at 487 with rmsd of 2.00382
WARNING - possible iterative failure at 488 with rmsd of 2.00083
WARNING - possible iterative failure at 695 with rmsd of 2.00148
WARNING - possible iterative failure at 810 with rmsd of 2.00007
WARNING - possible iterative failure at 826 with rmsd of 2.0013
WARNING - possible iterative failure at 832 with rmsd of 2.00119
WARNING - possible iterative failure at 835 with rmsd of 2.00006
WARNING - possible iterative failure at 909 with rmsd of 2.00041
WARNING - possible iterative failure at 912 with rmsd of 2.00373
WARNING - possible iterative failure at 981 with rmsd of 2.00328
*** There were 19 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************


-=-=-=-=-=-=-=- MACROCYCLE #3 -=-=-=-=-=-=-=-
**************************************************
Selected 28 atoms with selector 'name == 'CA''.
RMSD test = 1
Rotated rmsd = 36.1813
Aligned rmsd = 0.942025
[0.825472 -0.0168498 -0.564191 0.64941;0.0343746 0.9992 0.0204522 -2.37659;0.563395 -0.0362766 0.825391 3.00995;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 4
WARNING - possible iterative failure at 64 with rmsd of 2.00234
WARNING - possible iterative failure at 395 with rmsd of 2.00054
WARNING - possible iterative failure at 465 with rmsd of 2.00195
*** There were 3 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************
**************************************************
Selected 25 atoms with selector 'resid == 5'.
RMSD test = 1
Rotated rmsd = 22.1325
Aligned rmsd = 0.954096
[0.819819 0.099482 -0.563915 -2.47165;-0.00567425 0.986156 0.165722 0.90101;0.572594 -0.132662 0.809034 3.25036;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 6
WARNING - possible iterative failure at 694 with rmsd of 2.01215
WARNING - possible iterative failure at 903 with rmsd of 2.00008
WARNING - possible iterative failure at 983 with rmsd of 2.00795
*** There were 3 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************
**************************************************
Selected 125 atoms with selector 'segid == 'L2''.
RMSD test = 1
Rotated rmsd = 17.3223
Aligned rmsd = 0.994665
[0.811897 -0.00867864 -0.583736 -0.00435822;0.00763907 0.999962 -0.00424193 0.223086;0.58375 -0.00101519 0.811933 0.335095;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 4
WARNING - possible iterative failure at 56 with rmsd of 2.00257
WARNING - possible iterative failure at 85 with rmsd of 2.00126
WARNING - possible iterative failure at 191 with rmsd of 2.00133
WARNING - possible iterative failure at 221 with rmsd of 2.00216
WARNING - possible iterative failure at 254 with rmsd of 2.00138
WARNING - possible iterative failure at 407 with rmsd of 2.0023
WARNING - possible iterative failure at 581 with rmsd of 2.00192
*** There were 7 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************


-=-=-=-=-=-=-=- MACROCYCLE #4 -=-=-=-=-=-=-=-
**************************************************
Selected 28 atoms with selector 'name == 'CA''.
RMSD test = 1
Rotated rmsd = 36.3903
Aligned rmsd = 0.956309
[0.820947 -0.00720632 -0.570959 0.199285;-0.0197414 0.998964 -0.0409933 2.14853;0.570664 0.0449249 0.819954 -3.03994;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 4
WARNING - possible iterative failure at 174 with rmsd of 2.00037
*** There were 1 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************
**************************************************
Selected 25 atoms with selector 'resid == 5'.
RMSD test = 1
Rotated rmsd = 21.9677
Aligned rmsd = 0.921379
[0.781423 -0.000660299 -0.624001 -1.91322;-0.00977393 0.999864 -0.0132977 -0.5538;0.623925 0.0164901 0.781311 1.10769;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 6
WARNING - possible iterative failure at 247 with rmsd of 2.01963
WARNING - possible iterative failure at 653 with rmsd of 2.00501
WARNING - possible iterative failure at 737 with rmsd of 2.00001
*** There were 3 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************
**************************************************
Selected 125 atoms with selector 'segid == 'L2''.
RMSD test = 1
Rotated rmsd = 17.347
Aligned rmsd = 0.992599
[0.818593 0.00245938 -0.574368 -0.137206;-0.00377091 0.999992 -0.00109247 -0.0643601;0.574361 0.00306018 0.818596 -0.153972;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 4
WARNING - possible iterative failure at 187 with rmsd of 2.00293
WARNING - possible iterative failure at 280 with rmsd of 2.00073
WARNING - possible iterative failure at 653 with rmsd of 2.00218
WARNING - possible iterative failure at 685 with rmsd of 2.00286
WARNING - possible iterative failure at 821 with rmsd of 2.00123
*** There were 5 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************
